### DISCUSSION TITLE
Cannot type return value with new pipe syntax because of UnionType error

### DISCUSSION THREAD

**Discussion by AAraKKe** [microsoft/pyright#7258] [Q&A]
Hi, 

I have spent some time reading issues and discussions here concerning the new (I think) approach in pyright that is causing some problems with UnionType (and others).

I am sorry if this is around the same issue. Many of them revolve around using UnionType in generics, but this is a very simple usecase. I am failing to see how this behavior is possible. I have this method:

```python
def get_settings_from_user(cls, user_id: int) -> Settings | None:
    ...
```

This is a perfectly valid method annotation. A method that can either return the Settings object for the user or None if the user with such ID is not found. Typing the method like this generates a complaint from Pylance in VSCode (using pyright) that states

```
Expected type expression but received "UnionType"
```

Maybe this is obvious but I having a hard time understanding why this a wrong type annotation and pyright prefers `Optional[Settings]` which does not generate any issue.

Mypy does not complain about this and accepts the pipe annotation.

Is this an issue with understanding the `Settings | None` part?

**Comment by erictraut** [comment]
Could you provide a minimal, self-contained code sample? I can't glean enough details from the above code. For example, how is `Settings` defined? Is it a class? Is `get_settings_from_user` a class method?

**Reply by AAraKKe** [reply]
Hi Eric, sure. I am trying to be as close as possible to the source code. These classes are descendants of SQLModel; this is a self-contained piece of code where VSCode will complain about the `UnionType.` It is not executable though as there are some configurations for the db that I am leaving out but it can be validated by pyright.

```python
from sqlmodel import Field, Relationship, Session, SQLModel, create_engine, select


class Settings(SQLModel, table=True):
    id: int | None = Field(default=None, primary_key=True)
    user_id: int | None = Field(default=None, foreign_key="users.id")
    languaje: str = "en"
    timezone: str = "UTC"

    user: "User" = Relationship(back_populates="settings")


class User(SQLModel, table=True):
    first_name: str
    id: int | None = Field(default=None, primary_key=True)
    last_name: str | None = None
    username: str | None = None
    settings: Settings = Relationship(back_populates="user", sa_relationship_kwargs={"uselist": False})

    @classmethod
    def get_settings_from_user(cls, id: int) -> Settings | None:
        with Session(create_engine("url")) as session:
            statement = select(cls).where(cls.id == id)
            found_user = session.exec(statement).first()

            return None if found_user is None else found_user.settings
```

If I change the method definition to this one

```python
def get_settings_from_user(cls, id: int) -> Optional[Settings]:
```

pyright is happy with it. But this violates the linting and formatting rules we try to follow in the project using newer typing notations.

In general, VSCode is also having issues understanding the types of `user.settings.<field>` and reports them as unknown, but I am not sure if this is pyright and probably the language engine in VSCode.

**Reply by erictraut** [reply]
Thanks for the code sample. This looks like a bug. I've narrowed it down to a minimal code sample and filed a [bug report](https://github.com/microsoft/pyright/issues/7260).

**Reply by AAraKKe** [reply]
Thanks, Eric. May I ask what the `Unknown` type you refer to in the bug report? I find it odd that in the example above if I do something like `user.settings.timezone,` VSCode behaves very weirdly. The `settings` attribute appears as `Unkown` and when I do `settings.timezone` it is as if VSCode is not able to recognize the type and it appears as `(function) timezone: Unkown`.

Is Pyright providing this information to VSCode for the type of information we see when we hover over the variable? Could the fix for the bug report created resolve this?

Thanks!

**Reply by erictraut** [reply]
The `sqlmodel` package defines a class `SQLModelMetaclass` that derives from a symbol `ModelMetaclass`, but `ModelMetaclass` cannot be resolved statically because the package is doing some dynamic checks based on the runtime version of pydantic that is installed. That means the type of `ModelMetclass` is implicitly `Any`. That condition is unusual, but it's not illegal. It appears that the bug is related to this condition. I'll need to dig into it more to fully diagnose it.

And yes, the information that you see in the "hover text" is provided by pyright if you're using the Pylance language server in VS Code.

**Reply by erictraut** [reply]
The problem here is that the metaclass derives from an unknown class, so pyright cannot determine whether the metaclass implements a `__or__` method, which overrides the normal behavior of the `|` operator. It current assumes it might. I've changed the logic to assume the opposite. Both are valid guesses, but most metaclasses don't overload `__or__` so it's probably best to make an assumption that aligns with the common case.

**Reply by AAraKKe** [reply]
Thanks Eric! I am guessing that the dynamic behavior of SQLModel is temporary to support pydantic v1 and v2, eventually they should move to pydantic v2 fully and the dynamic behaviour should go away, my guess.

Does this mean the issue should have been resolved already?
